File: 1872-stone-game-viii/1872-stone-game-viii.cpp
class Solution {
public:
  
    int stoneGameVIII(vector<int>& stones) {
        //dp.resize(100005,vector<int>(2,INT_MIN));
        int n=stones.size();
        if(n==2)
        {
            return stones[0]+stones[1];
        }
        for(int i=1;i<n;i++)
        {
            stones[i]+=stones[i-1];
        }
        int dp[n];
        dp[n-1]=stones[n-1];
        int ans=dp[n-1];
        for(int i=n-2;i>=0;i--)
        {
            dp[i]=max(dp[i+1],stones[i]-dp[i+1]);
          if(i>0)
          {
                ans=max(ans,dp[i]);
          }
        }
           return ans;
    }
};

File: 1872-stone-game-viii/1872-stone-game-viii.java
         class Solution {
                    Integer[][] dp;
                    public int stoneGameVIII(int[] stones) {
                        int n = stones.length;
                        for(int i = 1; i < n; i++)
                            stones[i] += stones[i - 1];
                        dp = new Integer[n][2];
                        return recur(1, 0, stones);
                    }
                    
                    int recur(int idx, int turn, int[] prefix) {
                        // checking if we have already computed the result for current subproblem
                        // if yes, we will return the answer directly without further computations
                        // if no, we will compute the answer and store it as given below
                        if(dp[idx][turn] != null)
                            return dp[idx][turn];

                        int curScore = prefix[idx];

                        if(idx == prefix.length - 1) {
                            return turn == 0 ? curScore : -curScore;
                        }

                        if(turn % 2 == 0) {
                            // storing the result for current subproblem in DP array
                            dp[idx][turn] = Math.max(recur(idx + 1, turn, prefix), curScore + recur(idx + 1, turn ^ 1, prefix));
                        } else {
                            // storing the result for current subproblem in DP array
                            dp[idx][turn] = Math.min(recur(idx + 1, turn, prefix), -curScore + recur(idx + 1, turn ^ 1, prefix));
                        }

                        return dp[idx][turn];
                    }
                };